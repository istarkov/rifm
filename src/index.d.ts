import * as React from 'react';

interface RifmRenderArgs<E> {
  value: string;
  onChange: React.ChangeEventHandler<E>;
}

interface RifmProps<E> {
  value: string;
  onChange: (str: string) => void;
  format: (str: string) => string;
  replace?: (str: string) => string;
  append?: (str: string) => string;
  mask?: boolean;
  accept?: RegExp;
  children: (args: RifmRenderArgs<E>) => React.ReactNode;
}

declare class Rifm<E = HTMLInputElement> extends React.Component<
  RifmProps<E>
> {}

export { Rifm };
