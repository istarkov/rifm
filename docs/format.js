/* @flow */
/* simple fomtatters without number overflow checks etc */

import { AsYouType } from 'libphonenumber-js';

export const numberFormat = (str: string) => {
  const r = parseInt(str.replace(/[^\d]+/gi, ''), 10);

  return r ? r.toLocaleString('en') : '';
};

export const negNumberFormat = (str: string) => {
  const clean = str.replace(/[^\d-]+/gi, '');
  if (clean === '-') return '-';
  const r = parseInt(clean, 10);

  return r ? r.toLocaleString('en') : '';
};

export const currencyFormat = (str: string, isInitial?: boolean) => {
  const clean = str.replace(/[^\d.]+/gi, '');

  const beautify =
    clean.indexOf('.') === -1
      ? clean.length > 2 && isInitial !== true
        ? `${clean.substr(0, clean.length - 2)}.${clean.substr(-2)}`
        : clean
      : `${clean.split('.')[0]}.${clean.split('.')[1].substr(0, 2)}`;

  const r = parseFloat(beautify);

  return !Number.isNaN(r)
    ? r.toLocaleString('de-CH', {
        minimumFractionDigits: 2,
        maximumFractionDigits: 2,
      })
    : '';
};

const prepareFraction = (fraction, maximumFractionDigits) => {
  const removeTrailingZero = false;
  const substrFrac = fraction.substr(
    0,
    maximumFractionDigits -
      (removeTrailingZero && fraction[maximumFractionDigits - 1] === '0'
        ? 1
        : 0)
  );

  return substrFrac;
};

// poor raw prototype for future universal number format
export const currencyFormat2 = (str: string, isInitial?: boolean) => {
  const maximumFractionDigits = 2;
  const minimumFractionDigits = 0;

  const clean = str.replace(/[^\d.]+/gi, ''); // .replace(/0+$/, '');

  const [base, fract = ''] = clean.split('.');
  const fraction = prepareFraction(fract, maximumFractionDigits);
  const preventRounding = !fraction
    ? clean.length > minimumFractionDigits &&
      minimumFractionDigits > 0 &&
      isInitial !== true
      ? `${clean.substr(
          0,
          clean.length - minimumFractionDigits
        )}.${clean.substr(-minimumFractionDigits)}`
      : clean
    : `${base}.${fraction}`;

  const r = parseFloat(preventRounding);

  const formatted = r.toLocaleString('de-CH', {
    minimumFractionDigits,
    maximumFractionDigits,
  });
  const [, formattedFraction = ''] = formatted.split('.');

  const res = Number.isNaN(r)
    ? ''
    : formatted +
      (isInitial === true
        ? ''
        : (formattedFraction.length === 0 && clean.indexOf('.') > -1
            ? '.'
            : '') +
          (formattedFraction.length < maximumFractionDigits &&
          fraction.length > formattedFraction.length
            ? fraction.substring(
                formattedFraction.length,
                maximumFractionDigits
              )
            : ''));

  return res;
};

export const dateFormat = (str: string) => {
  const clean = str.replace(/[^\d]+/gi, '');
  const chars = clean.split('');
  return chars.reduce(
    (r, v, index) =>
      (index === 2 || index === 4 ? `${r}-${v}` : `${r}${v}`).substr(0, 10),
    ''
  );
};

const dtSym = '__-__-____';

export const dateFormatSym = (str: string) => {
  const clean = str.replace(/[^\d]+/gi, '');
  const chars = clean.split('');
  const r = chars.reduce(
    (r, v, index) =>
      (index === 2 || index === 4 ? `${r}-${v}` : `${r}${v}`).substr(0, 10),
    ''
  );

  if (r.length === 0) return r;

  return `${r}${dtSym.substr(r.length)}`;
};

const usPhone = new AsYouType('US');

export const formatPhone = (str: string) => {
  const clean = str.replace(/[^\d]+/gi, '').substr(0, 10);
  const r = usPhone.input(clean);
  usPhone.reset();

  return r;
};
