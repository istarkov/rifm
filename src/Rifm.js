/* @flow */

import * as React from 'react';

type Props = {|
  value: string,
  onChange: string => void,
  format: (str: string) => string,
  replace?: string => boolean,
  refuse?: RegExp,
  children: ({
    value: string,
    onChange: (
      evt: SyntheticInputEvent<HTMLInputElement | HTMLTextAreaElement>
    ) => void,
  }) => React.Node,
|};

export const Rifm = (props: Props) => {
  const [, refresh] = React.useReducer(c => c + 1, 0);
  const valueRef = React.useRef(null);

  // state of delete button see comments below about inputType support
  const isDeleleteButtonDownRef = React.useRef(false);

  const onChange = (
    evt: SyntheticInputEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    if (process.env.NODE_ENV !== 'production') {
      if (evt.target.type === 'number') {
        console.error(
          'Rifm does not support input type=number, use type=tel instead.'
        );
        return;
      }
    }

    const value = props.value;
    const eventValue = evt.target.value;

    valueRef.current = [
      eventValue, // eventValue
      evt.target, // input
      eventValue.length > value.length, // isSizeIncreaseOperation
      isDeleleteButtonDownRef.current, // isDeleleteButtonDown
      value === props.format(eventValue), // isNoOperation
    ];

    // The main trick is to update underlying input with non formatted value (= eventValue)
    // that allows us to calculate right cursor position after formatting (see getCursorPosition)
    // then we format new value and call props.onChange with masked/formatted value
    // and finally we are able to set cursor position into right place
    refresh();
  };

  // React prints warn on server in non production mode about useLayoutEffect usage
  // in both cases it's noop
  if (process.env.NODE_ENV === 'production' || typeof window !== 'undefined') {
    React.useLayoutEffect(() => {
      if (valueRef.current != null) {
        let [
          eventValue,
          input,
          isSizeIncreaseOperation,
          isDeleleteButtonDown,
          // No operation means that value itself hasn't been changed, BTW cursor, selection etc can be changed
          isNoOperation,
        ] = valueRef.current;
        valueRef.current = null;

        // this usually occurs on deleting special symbols like ' here 123'123.00
        // in case of isDeleleteButtonDown cursor should move differently vs backspace
        const deleteWasNoOp = isDeleleteButtonDown && isNoOperation;

        const refuse = props.refuse || /[^\d]+/g;

        const valueBeforeSelectionStart = eventValue
          .substr(0, input.selectionStart)
          .replace(refuse, '')
          .toLowerCase();

        // trying to find cursor position in formatted value having knowledge about valueBeforeSelectionStart
        // This works because we assume that format doesn't change the order of non refused symbols.
        // Imagine we have formatter which adds ' symbol between numbers, and by default we refuse all non numeric symbols
        // for example we had input = 1'2|'4 (| means cursor position) then user entered '3' symbol
        // inputValue = 1'23'|4 so valueBeforeSelectionStart = 123 and formatted value = 1'2'3'4
        // calling getCursorPosition("1'2'3'4") will give us position after 3, 1'2'3|'4
        // so for formatting just this function to determine cursor position after formatting is enough
        // with masking we need to do some additional checks see `replace` below
        const getCursorPosition = val => {
          let start = 0;
          for (let i = 0; i !== valueBeforeSelectionStart.length; ++i) {
            start = Math.max(
              start,
              val.toLowerCase().indexOf(valueBeforeSelectionStart[i], start) + 1
            );
          }
          return start;
        };

        // Masking part, for masks if size of mask is above some value (props.replace checks that)
        // we need to replace symbols instead of do nothing as like in format
        if (
          props.replace &&
          props.replace(props.value) &&
          isSizeIncreaseOperation &&
          !isNoOperation
        ) {
          let start = getCursorPosition(eventValue);

          const c = eventValue.substr(start).replace(refuse, '')[0];
          start = eventValue.indexOf(c, start);

          eventValue = `${eventValue.substr(0, start)}${eventValue.substr(
            start + 1
          )}`;
        }

        const formattedValue = props.format(eventValue);

        if (props.value === formattedValue) {
          // if nothing changed for formatted value, just refresh so props.value will be used at render
          refresh();
        } else {
          props.onChange(formattedValue);
        }

        return () => {
          let start = getCursorPosition(formattedValue);

          // Visually improves working with masked values,
          // like cursor jumping over refused symbols
          if (
            props.replace &&
            (isSizeIncreaseOperation ||
              (isDeleleteButtonDown && !deleteWasNoOp))
          ) {
            while (
              formattedValue[start] &&
              refuse.test(formattedValue[start])
            ) {
              start += 1;
            }
          }

          input.selectionStart = input.selectionEnd =
            start + (deleteWasNoOp ? 1 : 0);
        };
      }
    });
  }

  React.useEffect(() => {
    // until https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/inputType will be supported
    // by all major browsers (now supported by: +chrome, +safari, ?edge, !firefox)
    // there is no way I found to distinguish in onChange
    // backspace or delete was called in some situations
    // firefox track https://bugzilla.mozilla.org/show_bug.cgi?id=1447239
    const handleKeyDown = (evt: KeyboardEvent) => {
      if (evt.code === 'Delete') {
        isDeleleteButtonDownRef.current = true;
      }
    };

    const handleKeyUp = (evt: KeyboardEvent) => {
      if (evt.code === 'Delete') {
        isDeleleteButtonDownRef.current = false;
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    document.addEventListener('keyup', handleKeyUp);

    return () => {
      document.removeEventListener('keydown', handleKeyDown);
      document.removeEventListener('keyup', handleKeyUp);
    };
  }, []);

  return props.children({
    value: valueRef.current != null ? valueRef.current[0] : props.value,
    onChange,
  });
};
