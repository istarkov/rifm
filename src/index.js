/* @flow */

export { Rifm } from './rifm';
