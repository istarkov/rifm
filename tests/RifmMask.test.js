// @flow

import * as React from 'react';
import TestRenderer, { act } from 'react-test-renderer';
import { Value } from 'react-powerplug';
import { Rifm } from '../src';
import { dateFormat } from '../docs/format.js';
import {
  InputEmulator,
  renderInputState,
  type InputCommand,
} from './utils/InputEmulator';

test('mask behaviour', async () => {
  const snaphot = [];
  let getVal = null;
  let execCommand = null;
  let stateValue_ = null;

  TestRenderer.create(
    <Value initial={''}>
      {input => {
        stateValue_ = input.value;

        return (
          <Rifm
            replace={v => v.length >= 10}
            value={input.value}
            onChange={input.set}
            format={dateFormat}
          >
            {({ value, onChange }) => (
              <InputEmulator value={value} onChange={onChange}>
                {(exec, val) => {
                  execCommand = exec;
                  getVal = val;
                  return null;
                }}
              </InputEmulator>
            )}
          </Rifm>
        );
      }}
    </Value>
  );

  const exec = (cmd: InputCommand) => {
    if (!execCommand || !getVal) {
      throw Error('rifm is not initialized');
    }

    act(() => {
      if (!execCommand) {
        throw Error('rifm is not initialized');
      }

      execCommand(cmd);
    });

    expect(stateValue_).toEqual(getVal().value);
    snaphot.push({ ...getVal(), cmd, withCaret: renderInputState(getVal()) });
  };

  exec({ type: 'PUT_SYMBOL', payload: '1' });
  exec({ type: 'PUT_SYMBOL', payload: '23' });
  exec({ type: 'MOVE_CARET', payload: -1 });
  exec({ type: 'PUT_SYMBOL', payload: '4' });
  exec({ type: 'MOVE_CARET', payload: -100 }); // -100 at begin

  exec({ type: 'PUT_SYMBOL', payload: '5' });
  exec({ type: 'PUT_SYMBOL', payload: '6' });
  exec({ type: 'MOVE_CARET', payload: +100 }); // 100 at end
  exec({ type: 'PUT_SYMBOL', payload: '789' });
  // now check that replace works
  exec({ type: 'MOVE_CARET', payload: -4 });
  exec({ type: 'PUT_SYMBOL', payload: '9' });
  exec({ type: 'PUT_SYMBOL', payload: '8' });
  exec({ type: 'PUT_SYMBOL', payload: '7' });
  exec({ type: 'PUT_SYMBOL', payload: '6' });

  exec({ type: 'BACKSPACE' });
  exec({ type: 'PUT_SYMBOL', payload: '6' });

  exec({ type: 'MOVE_CARET', payload: -3 });
  exec({ type: 'BACKSPACE' });

  exec({ type: 'PUT_SYMBOL', payload: '0' });

  exec({ type: 'BACKSPACE' });
  exec({ type: 'PUT_SYMBOL', payload: '01' });

  exec({ type: 'PUT_SYMBOL', payload: '2345678' });

  exec({ type: 'MOVE_CARET', payload: -100 }); // -100 at begin
  exec({ type: 'MOVE_CARET', payload: 2 });

  exec({ type: 'BACKSPACE' });
  exec({ type: 'BACKSPACE' });

  exec({ type: 'PUT_SYMBOL', payload: '9876' });
  exec({ type: 'PUT_SYMBOL', payload: '5' });

  expect(snaphot).toMatchSnapshot();
});

test('mask behaviour with bad symbols', async () => {
  const snaphot = [];
  let getVal = null;
  let execCommand = null;
  let stateValue_ = null;

  TestRenderer.create(
    <Value initial={''}>
      {input => {
        stateValue_ = input.value;

        return (
          <Rifm
            replace={v => v.length >= 10}
            value={input.value}
            onChange={input.set}
            format={dateFormat}
          >
            {({ value, onChange }) => (
              <InputEmulator value={value} onChange={onChange}>
                {(exec, val) => {
                  execCommand = exec;
                  getVal = val;
                  return null;
                }}
              </InputEmulator>
            )}
          </Rifm>
        );
      }}
    </Value>
  );

  const exec = cmd => {
    if (!execCommand || !getVal) {
      throw Error('rifm is not initialized');
    }

    act(() => {
      if (!execCommand) {
        throw Error('rifm is not initialized');
      }

      execCommand(cmd);
    });

    expect(stateValue_).toEqual(getVal().value);
    snaphot.push({ ...getVal(), cmd, withCaret: renderInputState(getVal()) });
  };

  exec({ type: 'PUT_SYMBOL', payload: '18081978' });
  exec({ type: 'MOVE_CARET', payload: -4 });
  exec({ type: 'PUT_SYMBOL', payload: 'x' });

  expect(snaphot).toMatchSnapshot();
});

test('mask behaviour with delete', async () => {
  const snaphot = [];
  let getVal = null;
  let execCommand = null;
  let stateValue_ = null;

  TestRenderer.create(
    <Value initial={''}>
      {input => {
        stateValue_ = input.value;

        return (
          <Rifm
            replace={v => v.length >= 10}
            value={input.value}
            onChange={input.set}
            format={dateFormat}
          >
            {({ value, onChange }) => (
              <InputEmulator value={value} onChange={onChange}>
                {(exec, val) => {
                  execCommand = exec;
                  getVal = val;
                  return null;
                }}
              </InputEmulator>
            )}
          </Rifm>
        );
      }}
    </Value>
  );

  const exec = cmd => {
    if (!execCommand || !getVal) {
      throw Error('rifm is not initialized');
    }

    act(() => {
      if (!execCommand) {
        throw Error('rifm is not initialized');
      }

      execCommand(cmd);
    });

    expect(stateValue_).toEqual(getVal().value);
    snaphot.push({ ...getVal(), cmd, withCaret: renderInputState(getVal()) });
  };

  exec({ type: 'PUT_SYMBOL', payload: '18081978' });
  exec({ type: 'MOVE_CARET', payload: -4 });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });
  exec({ type: 'PUT_SYMBOL', payload: '1978' });
  exec({ type: 'MOVE_CARET', payload: -100 });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });
  exec({ type: 'PUT_SYMBOL', payload: '18' });
  exec({ type: 'MOVE_CARET', payload: 1 });
  exec({ type: 'DELETE' });
  exec({ type: 'PUT_SYMBOL', payload: '8' });
  exec({ type: 'MOVE_CARET', payload: 2 });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });
  exec({ type: 'DELETE' });
  exec({ type: 'PUT_SYMBOL', payload: '78' });
  exec({ type: 'MOVE_CARET', payload: -5 });
  exec({ type: 'DELETE' });

  expect(snaphot).toMatchSnapshot();
});

test('mask works even if state is not updated on equal vals', async () => {
  const snaphot = [];
  let getVal = null;
  let execCommand = null;
  let stateValue_ = null;
  let callCount_ = 0;

  TestRenderer.create(
    <Value initial={''}>
      {input => {
        stateValue_ = input.value;

        return (
          <Rifm
            replace={v => v.length >= 10}
            value={input.value}
            onChange={v => {
              if (input.value !== v) {
                callCount_++;
                input.set(v);
              }
            }}
            format={dateFormat}
          >
            {({ value, onChange }) => (
              <InputEmulator value={value} onChange={onChange}>
                {(exec, val) => {
                  execCommand = exec;
                  getVal = val;
                  return null;
                }}
              </InputEmulator>
            )}
          </Rifm>
        );
      }}
    </Value>
  );

  const exec = cmd => {
    if (!execCommand || !getVal) {
      throw Error('rifm is not initialized');
    }

    act(() => {
      if (!execCommand) {
        throw Error('rifm is not initialized');
      }

      execCommand(cmd);
    });

    expect(stateValue_).toEqual(getVal().value);
    snaphot.push({ ...getVal(), cmd, withCaret: renderInputState(getVal()) });
  };

  exec({ type: 'PUT_SYMBOL', payload: '18081978' });
  exec({ type: 'MOVE_CARET', payload: -100 });
  exec({ type: 'PUT_SYMBOL', payload: '18081978' });
  exec({ type: 'MOVE_CARET', payload: -4 });
  exec({ type: 'PUT_SYMBOL', payload: '1978' });
  exec({ type: 'MOVE_CARET', payload: -4 });
  exec({ type: 'BACKSPACE' });
  exec({ type: 'DELETE' });

  exec({ type: 'PUT_SYMBOL', payload: 'x' });
  exec({ type: 'MOVE_CARET', payload: -1 });
  exec({ type: 'PUT_SYMBOL', payload: 'x' });
  exec({ type: 'MOVE_CARET', payload: -1 });
  exec({ type: 'PUT_SYMBOL', payload: '1' });

  expect(callCount_).toEqual(1);

  expect(snaphot).toMatchSnapshot();
});
