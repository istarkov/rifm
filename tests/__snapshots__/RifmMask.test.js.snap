// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mask behaviour 1`] = `
Array [
  Object {
    "cmd": Object {
      "payload": "1",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "1",
    "withCaret": "1|",
  },
  Object {
    "cmd": Object {
      "payload": "23",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "12-3",
    "withCaret": "12-3|",
  },
  Object {
    "cmd": Object {
      "payload": -1,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 3,
    "selectionStart": 3,
    "value": "12-3",
    "withCaret": "12-|3",
  },
  Object {
    "cmd": Object {
      "payload": "4",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "12-43",
    "withCaret": "12-4|3",
  },
  Object {
    "cmd": Object {
      "payload": -100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "12-43",
    "withCaret": "|12-43",
  },
  Object {
    "cmd": Object {
      "payload": "5",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "51-24-3",
    "withCaret": "5|1-24-3",
  },
  Object {
    "cmd": Object {
      "payload": "6",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 3,
    "selectionStart": 3,
    "value": "56-12-43",
    "withCaret": "56-|12-43",
  },
  Object {
    "cmd": Object {
      "payload": 100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 8,
    "selectionStart": 8,
    "value": "56-12-43",
    "withCaret": "56-12-43|",
  },
  Object {
    "cmd": Object {
      "payload": "789",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 10,
    "selectionStart": 10,
    "value": "56-12-4378",
    "withCaret": "56-12-4378|",
  },
  Object {
    "cmd": Object {
      "payload": -4,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "56-12-4378",
    "withCaret": "56-12-|4378",
  },
  Object {
    "cmd": Object {
      "payload": "9",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 7,
    "selectionStart": 7,
    "value": "56-12-9378",
    "withCaret": "56-12-9|378",
  },
  Object {
    "cmd": Object {
      "payload": "8",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 8,
    "selectionStart": 8,
    "value": "56-12-9878",
    "withCaret": "56-12-98|78",
  },
  Object {
    "cmd": Object {
      "payload": "7",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 9,
    "selectionStart": 9,
    "value": "56-12-9878",
    "withCaret": "56-12-987|8",
  },
  Object {
    "cmd": Object {
      "payload": "6",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 10,
    "selectionStart": 10,
    "value": "56-12-9876",
    "withCaret": "56-12-9876|",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 9,
    "selectionStart": 9,
    "value": "56-12-987",
    "withCaret": "56-12-987|",
  },
  Object {
    "cmd": Object {
      "payload": "6",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 10,
    "selectionStart": 10,
    "value": "56-12-9876",
    "withCaret": "56-12-9876|",
  },
  Object {
    "cmd": Object {
      "payload": -3,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 7,
    "selectionStart": 7,
    "value": "56-12-9876",
    "withCaret": "56-12-9|876",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 5,
    "selectionStart": 5,
    "value": "56-12-876",
    "withCaret": "56-12|-876",
  },
  Object {
    "cmd": Object {
      "payload": "0",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 7,
    "selectionStart": 7,
    "value": "56-12-0876",
    "withCaret": "56-12-0|876",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 5,
    "selectionStart": 5,
    "value": "56-12-876",
    "withCaret": "56-12|-876",
  },
  Object {
    "cmd": Object {
      "payload": "01",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 8,
    "selectionStart": 8,
    "value": "56-12-0187",
    "withCaret": "56-12-01|87",
  },
  Object {
    "cmd": Object {
      "payload": "2345678",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 10,
    "selectionStart": 10,
    "value": "56-12-0123",
    "withCaret": "56-12-0123|",
  },
  Object {
    "cmd": Object {
      "payload": -100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "56-12-0123",
    "withCaret": "|56-12-0123",
  },
  Object {
    "cmd": Object {
      "payload": 2,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 2,
    "selectionStart": 2,
    "value": "56-12-0123",
    "withCaret": "56|-12-0123",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 1,
    "selectionStart": 1,
    "value": "51-20-123",
    "withCaret": "5|1-20-123",
  },
  Object {
    "cmd": Object {
      "type": "BACKSPACE",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "12-01-23",
    "withCaret": "|12-01-23",
  },
  Object {
    "cmd": Object {
      "payload": "9876",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "98-76-1201",
    "withCaret": "98-76-|1201",
  },
  Object {
    "cmd": Object {
      "payload": "5",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 7,
    "selectionStart": 7,
    "value": "98-76-5201",
    "withCaret": "98-76-5|201",
  },
]
`;

exports[`mask behaviour with bad symbols 1`] = `
Array [
  Object {
    "cmd": Object {
      "payload": "18081978",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 10,
    "selectionStart": 10,
    "value": "18-08-1978",
    "withCaret": "18-08-1978|",
  },
  Object {
    "cmd": Object {
      "payload": -4,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "18-08-1978",
    "withCaret": "18-08-|1978",
  },
  Object {
    "cmd": Object {
      "payload": "x",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "18-08-1978",
    "withCaret": "18-08-|1978",
  },
]
`;

exports[`mask behaviour with delete 1`] = `
Array [
  Object {
    "cmd": Object {
      "payload": "18081978",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 10,
    "selectionStart": 10,
    "value": "18-08-1978",
    "withCaret": "18-08-1978|",
  },
  Object {
    "cmd": Object {
      "payload": -4,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "18-08-1978",
    "withCaret": "18-08-|1978",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "18-08-978",
    "withCaret": "18-08-|978",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "18-08-78",
    "withCaret": "18-08-|78",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "18-08-8",
    "withCaret": "18-08-|8",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 5,
    "selectionStart": 5,
    "value": "18-08",
    "withCaret": "18-08|",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "18-08",
    "withCaret": "18-08|",
  },
  Object {
    "cmd": Object {
      "payload": "1978",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 10,
    "selectionStart": 10,
    "value": "18-08-1978",
    "withCaret": "18-08-1978|",
  },
  Object {
    "cmd": Object {
      "payload": -100,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "18-08-1978",
    "withCaret": "|18-08-1978",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "80-81-978",
    "withCaret": "|80-81-978",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 0,
    "selectionStart": 0,
    "value": "08-19-78",
    "withCaret": "|08-19-78",
  },
  Object {
    "cmd": Object {
      "payload": "18",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 3,
    "selectionStart": 3,
    "value": "18-08-1978",
    "withCaret": "18-|08-1978",
  },
  Object {
    "cmd": Object {
      "payload": 1,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "18-08-1978",
    "withCaret": "18-0|8-1978",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 4,
    "selectionStart": 4,
    "value": "18-01-978",
    "withCaret": "18-0|1-978",
  },
  Object {
    "cmd": Object {
      "payload": "8",
      "type": "PUT_SYMBOL",
    },
    "selectionEnd": 6,
    "selectionStart": 6,
    "value": "18-08-1978",
    "withCaret": "18-08-|1978",
  },
  Object {
    "cmd": Object {
      "payload": 2,
      "type": "MOVE_CARET",
    },
    "selectionEnd": 8,
    "selectionStart": 8,
    "value": "18-08-1978",
    "withCaret": "18-08-19|78",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 8,
    "selectionStart": 8,
    "value": "18-08-198",
    "withCaret": "18-08-19|8",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 8,
    "selectionStart": 8,
    "value": "18-08-19",
    "withCaret": "18-08-19|",
  },
  Object {
    "cmd": Object {
      "type": "DELETE",
    },
    "selectionEnd": 9,
    "selectionStart": 9,
    "value": "18-08-19",
    "withCaret": "18-08-19|",
  },
]
`;
